docs/design.md

```diff
@@ -0,0 +1,59 @@
+# Spark Operation Design
+
+Table of Contents
+=================
+- [Spark Operation Design](#spark-operation-design)
+- [Table of Contents](#table-of-contents)
+    - [Introduction](#introduction)
+    - [Why Spark Operator?](#why-spark-operator)
+    - [Components](#components)
+        - [Controller, Submission Runner, and Spark Pod Monitor](#controller-submission-runner-and-spark-pod-monitor)
+        - [Spark Pod Initializer](#spark-pod-initializer)
+        - [Command-line Tool: Sparkctl](#command-line-tool-sparkctl)
+
+## Introduction
+
+The Spark Operator follows the recent trend of leveraging the [operator](https://coreos.com/blog/introducing-operators.html) pattern for managing the life cycle of Spark applications on a Kubernetes cluster. The Spark Operator allows Spark applications to be specified in a declarative manner (e.g., in a YAML file) and run without the need to deal with the spark submission process. It also enables status of Spark applications to be tracked and presented idiomatically like other types of workloads on Kubernetes. This document discusses the design and architecture of the Spark Operator. For documentation of the [CustomResourceDefinition](https://kubernetes.io/docs/concepts/api-extension/custom-resources/) for specification of Spark applications, please refer to [API Definition](api.md)    
+
+## Why Spark Operator?
+
+## Components
+
+The Spark Operator consists of:
+* a `SparkApplication` controller that watches events of creation, updates, and deletion of 
+`SparkApplication` objects and acts on the watch events,
+* a *submission runner* that runs `spark-submit` for submissions received from the controller,
+* a *Spark pod monitor* that watches for Spark pods and sends pod status updates to the controller,
+* a Spark pod [initializer](https://kubernetes.io/docs/admin/extensible-admission-controllers/#initializers)that performs initialization tasks on Spark driver and executor pods based on the annotations on the pods added by the controller,
+* and also a command-line tool named `sparkctl` for working with the operator. 
+
+The following diagram shows how different components interact and work together.
+
+![Architecture Diagram](architecture-diagram.png)
+
+Specifically, a user uses the `sparkctl` (or `kubectl`) to create a `SparkApplication` object. The `SparkApplication` controller receives the object through a watcher from the API server, creates a submission carrying the `spark-submit` arguments, and sends the submission to the *submission runner*. The submission runner submits the application to run and creates the driver pod of the application. Upon starting, the driver pod creates the executor pods. While the application is running, the *Spark pod monitor* watches the pods of the application and sends status updates of the pods back to the controller, which then updates the status of the application accordingly. 
```

Comment:
When you talk about creating the driver Pod, and the driver creating executor Pods, are these really bare Pods? For example, what happens if a Pod gets preempted or drained?

Reply:
They are bare pods. The Spark submission client (called by `spark-submit`) creates the driver pod, which in turn creates the executor pods. We (the Spark k8s community) haven't really given much thoughts on pod preemption.  But this is a good point. I will need to do some investigation.

Reply:
I assume the driver would just make a replacement.  It has a reconciliation loop, right?

Reply:
No, there's not a reconciliation loop in Spark on K8s itself. The submission client simply creates the driver pod and watches the pod if it is configured to wait for the application to complete. If the pod completes, fails, or gets deleted, the client simply terminates and reports the exit code of the driver pod if it's available. This issue actually falls into the scope of high-availability driver, which is critical for Spark Streaming applications. Some people from Tencent addressed the failure case by using a K8s Job to run the driver. This is the only take on the issue as far as I am aware of.

Reply:
Yes, supporting restart with a configurable restart policy is a goal and a feature. Unlike pods in other types of workloads, particularly stateless workloads, the driver pod manages the executor pods and owns them (through `OwnerReference` on the executor pods). The Spark submission client is responsible for building and creating the driver pod. In this design, replacing the driver pod is done by invoking the submission client that re-creates a new driver pod.